plutus-chain-index-core/src/Plutus/ChainIndex/Tx.hs

```diff
@@ -130,30 +72,42 @@ txOutRefMapForAddr addr tx =
 -- 'OnChainTx' will be the inputs of the 'ChainIndexTx'.
 fromOnChainTx :: OnChainTx -> ChainIndexTx
 fromOnChainTx = \case
-    Valid tx@Tx{txInputs, txOutputs, txValidRange, txData, txMintScripts} ->
-        let (validatorHashes, otherDataHashes, redeemers) = validators txInputs in
-        ChainIndexTx
-            { _citxTxId = txId tx
-            , _citxInputs = txInputs
-            , _citxOutputs = ValidTx txOutputs
-            , _citxValidRange = txValidRange
-            , _citxData = txData <> otherDataHashes
-            , _citxRedeemers = redeemers
-            , _citxScripts = mintingPolicies txMintScripts <> validatorHashes
-            , _citxCardanoTx = Nothing
-            }
-    Invalid tx@Tx{txCollateral, txValidRange, txData, txInputs, txMintScripts} ->
-        let (validatorHashes, otherDataHashes, redeemers) = validators txInputs in
-        ChainIndexTx
-            { _citxTxId = txId tx
-            , _citxInputs = txCollateral
-            , _citxOutputs = InvalidTx
-            , _citxValidRange = txValidRange
-            , _citxData = txData <> otherDataHashes
-            , _citxRedeemers = redeemers
-            , _citxScripts = mintingPolicies txMintScripts <> validatorHashes
-            , _citxCardanoTx = Nothing
-            }
+    Valid ctx ->
+        onCardanoTx
+            (\tx@Tx{txInputs, txOutputs, txValidRange, txData, txMintScripts} ->
+                let (validatorHashes, otherDataHashes, redeemers) = validators txInputs in
+                ChainIndexTx
+                    { _citxTxId = txId tx
+                    , _citxInputs = txInputs
+                    , _citxOutputs = ValidTx txOutputs
+                    , _citxValidRange = txValidRange
+                    , _citxData = txData <> otherDataHashes
+                    , _citxRedeemers = redeemers
+                    , _citxScripts = mintingPolicies txMintScripts <> validatorHashes
+                    , _citxCardanoTx = Nothing
+                    })
+            (fromOnChainCardanoTx True)
+            ctx
+    Invalid ctx ->
+        onCardanoTx
+            (\tx@Tx{txCollateral, txValidRange, txData, txInputs, txMintScripts} ->
+                let (validatorHashes, otherDataHashes, redeemers) = validators txInputs in
+                ChainIndexTx
+                    { _citxTxId = txId tx
+                    , _citxInputs = txCollateral
+                    , _citxOutputs = InvalidTx
+                    , _citxValidRange = txValidRange
+                    , _citxData = txData <> otherDataHashes
+                    , _citxRedeemers = redeemers
+                    , _citxScripts = mintingPolicies txMintScripts <> validatorHashes
+                    , _citxCardanoTx = Nothing
+                    })
+            (fromOnChainCardanoTx False)
+            ctx
+
+fromOnChainCardanoTx :: Bool -> SomeCardanoApiTx -> ChainIndexTx
```

Comment:
Comment on why we need to manually set the script validity when converting from CardanoTx to ChainIndexTx?